### PR TITLE
refactor(assump) : refactor facts

### PR DESCRIPTION
**Pull request by JSS95**
#### References to other Issues or PRs

Suggested by @oscarbenjamin  in https://github.com/sympy/sympy/pull/21375#discussion_r624555911

#### Brief description of what is fixed or changed

New boolean function `Exclusive` is introduced. `Exclusive` returns `True` if only one or no argument is true. This is designed to be used in `assumptions` module to register the facts for the predicates.

#### Release Notes

NO ENTRY

### CODE REVIEW

File: sympy/assumptions/facts.py [hunk 1]
```diff
@@ -53,40 +54,38 @@ def get_known_facts(x=None):
...
+        Exclusive(Q.real(x), Q.imaginary(x)),
+        Implies(Q.real(x) | Q.imaginary(x), Q.complex(x)),
 
         # other subsets of complex
-        Implies(Q.transcendental(x), ~Q.algebraic(x)),
-        Implies(Q.irrational(x), ~Q.rational(x)),
-        Equivalent(Q.rational(x) | Q.irrational(x), Q.negative(x) | Q.zero(x) | Q.positive(x)),
+        Exclusive(Q.transcendental(x), Q.algebraic(x)),
+        Equivalent(Q.real(x), Q.rational(x) | Q.irrational(x)),
+        Exclusive(Q.irrational(x), Q.rational(x)),
```
Comment (oscarbenjamin) [inline sympy/assumptions/facts.py:70]: Do we have `Equivalent(complex, transcental | algebraic)` anywhere?

File: sympy/assumptions/facts.py [hunk 2]
```diff
@@ -53,40 +54,38 @@ def get_known_facts(x=None):
...
-        Implies(Q.negative(x) | Q.zero(x) | Q.positive(x), Q.hermitian(x)),
+        Implies(Q.real(x), Q.hermitian(x)),
         Implies(Q.imaginary(x), Q.antihermitian(x)),
         Implies(Q.zero(x), Q.hermitian(x) | Q.antihermitian(x)),
 
         # define finity and infinity, and build extended real line
-        Implies(Q.infinite(x), ~Q.finite(x)),
-        Implies(Q.algebraic(x) | Q.transcendental(x), Q.finite(x)),
+        Exclusive(Q.infinite(x), Q.finite(x)),
+        Equivalent(Q.complex(x), Q.finite(x)),
```
Comment (oscarbenjamin) [inline sympy/assumptions/facts.py:88]: I'm not sure about these being equivalent. This means that finite implies complex. We don't have that in the old assumptions.

File: sympy/logic/tests/test_boolalg.py [hunk 3]
```diff
@@ -227,6 +227,13 @@ def test_Equivalent():
     assert Equivalent(Equality(A, B), Equality(B, A)) is true
 
 
+def test_Exclusive():
+    assert Exclusive(False, False, False)
+    assert Exclusive(True, False, False)
+    assert not Exclusive(True, True, False)
+    assert not Exclusive(True, True, True)
```
Comment (oscarbenjamin) [inline sympy/logic/tests/test_boolalg.py:234]: It's better to use e.g. `is True` and `is False` in tests. Most objects `x` in Python will pass an `assert x`.

**Comment by oscarbenjamin** [comment]
I left comments but otherwise looks good.

File: sympy/assumptions/facts.py (hunk 1 above)
Comment (JSS95) [reply sympy/assumptions/facts.py:70]: Not explicitly, but `complex` is always converted to `algebraic | transcendental` by predicate decomposition.

Comment (JSS95) [reply sympy/assumptions/facts.py:70]: I'll mark this as resolved since this approach is approved when predicate decomposition was introduced.

**Pull request merged**